secboot/secboot_tpm.go

```diff
@@ -382,19 +400,32 @@ func unlockEncryptedPartitionWithSealedKey(tpm *sb.TPMConnection, name, device,
 		options.RecoveryKeyTries = 3
 	}
 
+	var method UnlockMethod
+
 	// XXX: pinfile is currently not used
 	activated, err := sbActivateVolumeWithTPMSealedKey(tpm, name, device, keyfile, nil, &options)
+
+	// with non-nil err, we should check for err being ActivateWithTPMSealedKeyError
+	// and RecoveryKeyUsageErr inside that being nil - this indicates that the
+	// recovery key was used to unlock it
 	if !activated {
 		// ActivateVolumeWithTPMSealedKey should always return an error if activated == false
-		return fmt.Errorf("cannot activate encrypted device %q: %v", device, err)
+		return NotUnlocked, fmt.Errorf("cannot activate encrypted device %q: %v", device, err)
 	}
 	if err != nil {
+		if activateErr, ok := err.(*sb.ActivateWithTPMSealedKeyError); ok {
+			if activateErr.RecoveryKeyUsageErr == nil {
+				// then we know for sure that it was unlocked with recover key
+				method = UnlockedWithRecoveryKey
+			}
+		}
```

Comment:
in the else path of this method stays 0, NotUnlocked, that seems strange ??

Reply:
Looking at the doc comment around ActivateVolumeWithTPMSealedKey unlocking with recovery key was only successful when RecoverykeyUsageErr == nil, otherwise the volume remains locked.